web/server.js

```diff
@@ -209,13 +232,25 @@ module.exports = function(options) {
         },
         pre: [
           {
+            assign: 'grant_type',
+            method: function (request, reply) {
+              reply(request.payload.grant_type);
+            }
+          },
+          {
             assign: 'client',
             method: function(request, reply) {
               oauthDb.getClient(request.payload.client_id, function(err, client) {
                 if ( err ) {
                   return reply(err);
                 }
-                if ( client.client_secret !== request.payload.client_secret ) {
+                if (
+                  client.allowed_grants.indexOf(request.pre.grant_type) === -1 ||
+                  (
+                    request.pre.grant_type === 'authorization_code' &&
```

Comment:
That would require removing all pre methods below into the handler. are we okay with that?